lib/spack/spack/main.py

```diff
@@ -389,8 +390,16 @@ def make_argument_parser(**kwargs):
     return parser
 
 
+def send_warning_to_tty(message, *args):
+    """Redirects messages to tty.warn."""
+    tty.warn(message)
+
+
 def setup_main_options(args):
     """Configure spack globals based on the basic options."""
+    # Assign a custom function to show warnings
+    warnings.showwarning = send_warning_to_tty
```

Comment:
Is this useful or is this going to make Spack even more verbose?  I think we should start to reduce the amount of noise users see, not increase it, so if this can be put into some message that does not go to the user, that is likely good.

Reply:
This shouldn't affect the verbosity of Spack. I decided to use [warnings](https://docs.python.org/3.7/library/warnings.html) so that `archspec` won't depend on `llnl.util.tty` to emit warnings. There are a couple of nice features that this package brings in:

1. There are filters that control whether a warning should issue a message or not
2. Issued messages are formatted and printed via a user settable hook

Above I used the second capability to let Spack display warnings via `llnl.util.tty`. Another nice feature is that by default a warning is emitted only one time per location - and the behavior is [customizable](https://docs.python.org/3.7/library/warnings.html#the-warnings-filter).

Reply:
oh nice -- that sounds great!